src/printer.js

```diff
@@ -3260,6 +3287,112 @@ function hasLeadingOwnLineComment(text, node) {
   return res;
 }
 
+function hasNakedLeftSide(node) {
+  return (
+    node.type === "BinaryExpression" ||
+    node.type === "LogicalExpression" ||
+    node.type === "ConditionalExpression" ||
+    node.type === "CallExpression" ||
+    node.type === "MemberExpression" ||
+    node.type === "SequenceExpression" ||
+    node.type === "TaggedTemplateExpression"
+  );
+}
+
+function getLeftSide(node) {
+  if (node.expressions) {
+    return node.expressions[0];
+  }
+  return node.left || node.test || node.callee || node.object || node.tag;
+}
+
+function exprNeedsASIProtection(node) {
+  const maybeASIProblem = node.needsParens ||
+    node.type === "ParenthesizedExpression" ||
+    node.type === "ArrayExpression" ||
+    (node.type === "UnaryExpression" &&
+      node.prefix &&
+      (node.operator === "+" || node.operator === "-")) ||
+    node.type === "TemplateLiteral" ||
+    node.type === "TemplateElement" ||
+    node.type === "JSXElement" ||
+    node.type === "RegExpLiteral" ||
+    (node.type === "Literal" && node.pattern) ||
+    (node.type === "Literal" && node.regex);
+
+  if (maybeASIProblem) {
+    return true;
+  }
+
+  if (!hasNakedLeftSide(node)) {
+    return false;
+  }
+
+  return exprNeedsASIProtection(getLeftSide(node));
+}
+
+function stmtNeedsASIProtection(path) {
+  if (!path) return false;
+  const node = path.getNode();
+
+  if (node.type !== "ExpressionStatement") {
+    return false;
+  }
+
+  return exprNeedsASIProtection(node.expression);
+}
+
+
+function classPropMayCauseASIProblems(path) {
+  const node = path.getNode();
+
+  if (node.type !== "ClassProperty") {
+    return false;
+  }
+
+  const name = node.key && node.key.name;
+  if (!name) {
+    return false;
+  }
+
+  // this isn't actually possible yet with most parsers available today
+  // so isn't properly tested yet.
```

Comment:
I did test this locally by changing `static` to `staticc` both here and in the tests, so I know it works, but I feel pretty iffy about leaving this in regardless... thoughts?